setup.py

```diff
@@ -11,35 +11,37 @@
 
 exec(compile(version_file_contents, "pymbolic/version.py", "exec"), ver_dic)
 
-setup(name="pymbolic",
-      version=ver_dic["VERSION_TEXT"],
-      description="A package for symbolic computation",
-      long_description=open("README.rst").read(),
-      classifiers=[
-          "Development Status :: 4 - Beta",
-          "Intended Audience :: Developers",
-          "Intended Audience :: Other Audience",
-          "Intended Audience :: Science/Research",
-          "License :: OSI Approved :: MIT License",
-          "Natural Language :: English",
-          "Programming Language :: Python",
-          "Programming Language :: Python :: 3",
-          "Topic :: Scientific/Engineering",
-          "Topic :: Scientific/Engineering :: Mathematics",
-          "Topic :: Software Development :: Libraries",
-          "Topic :: Utilities",
-          ],
-      author="Andreas Kloeckner",
-      author_email="inform@tiker.net",
-      license="MIT",
-      url="http://mathema.tician.de/software/pymbolic",
-
-      packages=find_packages(),
-      python_requires="~=3.8",
-      install_requires=[
-          "pytools>=2022.1.14",
-          ],
-      extras_require={
-          "test": ["pytest>=2.3"],
-          },
-      )
+setup(
+    name="pymbolic",
+    version=ver_dic["VERSION_TEXT"],
+    description="A package for symbolic computation",
+    long_description=open("README.rst").read(),
+    classifiers=[
+        "Development Status :: 4 - Beta",
+        "Intended Audience :: Developers",
+        "Intended Audience :: Other Audience",
+        "Intended Audience :: Science/Research",
+        "License :: OSI Approved :: MIT License",
+        "Natural Language :: English",
+        "Programming Language :: Python",
+        "Programming Language :: Python :: 3",
+        "Topic :: Scientific/Engineering",
+        "Topic :: Scientific/Engineering :: Mathematics",
+        "Topic :: Software Development :: Libraries",
+        "Topic :: Utilities",
+    ],
+    author="Andreas Kloeckner",
+    author_email="inform@tiker.net",
+    license="MIT",
+    url="http://mathema.tician.de/software/pymbolic",
+    packages=find_packages(),
+    python_requires="~=3.8",
+    install_requires=[
+        "pytools>=2022.1.14",
+        "attrs",
```

Comment:
I started out on `dataclasses`, but then switch to attrs for some constructor shenanigans that I ultimately didn't end up using. `attrs` has hash caching though, which I like. Outside of that, we could also switch to dataclasses without much of an issue.

Reply:
Going by what's implemented in `attrs` for caching, doing that ourselves seems like a bit of a pain too:
https://github.com/python-attrs/attrs/blob/46053d703d99a6e7419fd291cbf546340690d69d/src/attr/_make.py#L1620

Reply:
It doesn't quite have to be that gross. I implemented something simple, I think it suffices. `dataclasses` gets away with something super-simple, too: https://github.com/python/cpython/blob/77d25795862f19c6e3d647b76cfb10d5ce1f149c/Lib/dataclasses.py#L675-L680